File: src/services/champselect.ts
// 选人阶段分析：从 LCU session 数据 -> BP 推荐
import { getChampSelectSession, getMyTierId, getSummoner, queueToMode, isRankedMode, isHextechMode, type ChampSelectSession } from '../api/lcu.js';
import { getHeroList } from '../api/cn101.js';
import { recommendPick } from './pick.js';
import { recommendBan } from './ban.js';
import { recommendHextechHeroes } from './hextech.js';
import { type PickRecommendation, type BanRecommendation, type GameMode, TIER_NAMES, type TierId } from '../models.js';
import { normalizeLane } from './pick.js';

export interface ChampSelectAnalysis {
  phase: string;
  /** 模式（ranked_solo/ranked_flex/aram/hextech_aram/normal/other） */
  mode: GameMode;
  /** 模式中文名 */
  modeLabel: string;
  queueId?: number;
  /** 自己名字 */
  me: string;
  /** 对面已选（非锁定/已锁定英雄 id 列表） */
  enemyPicks: number[];
  /** 我方已选 */
  myPicks: number[];
  /** 双方 ban（含 0=未ban） */
  myBans: number[];
  enemyBans: number[];
  /** 我方剩余位置（assignedPosition） */
  myOpenLanes: string[];
  /** 我方 assignedPosition */
  myLane: string;
  /** 自己的英雄 id（海斗：翻牌/选定后 >0，用于展示该英雄的海斗榜推荐） */
  myHeroId: number;
  picks: PickRecommendation[];
  bans: BanRecommendation[];
  /** 海克斯/大乱斗英雄推荐（仅 aram 模式） */
  aramHeroes?: Awaited<ReturnType<typeof recommendHextechHeroes>>;
  /** 海克斯大乱斗共享池：我方翻开的英雄（进池后全员可选），按胜率排序（仅 hextech_aram） */
  aramPool: AramPoolEntry[];
  /** 生效段位 id（LCU 自动获取，默认 255 全段位） */
  tier: number;
  /** 生效段位名 */
  tierName: string;
  // ---------- 阶段感知（贴合真实 BP 流程） ----------
  /** timer 阶段中文：位置规划 / 禁用选择 / 最终确认 */
  timerPhase: string;
  /** 本阶段剩余秒数 */
  timeLeftSec: number;
  /** 当前进行中的操作（null = 全部完成/等待） */
  currentAction: {
    type: 'ban' | 'pick';
    /** 操作者名字 */
    actorName: string;
    /** 是否我方 */
    isAlly: boolean;
    /** 是否是我 */
    isMe: boolean;
    /** 操作者位置（LCU 口径） */
    lane: string;
  } | null;
  /** 已完成操作数（第几手，1-10） */
  completedActions: number;
  /** 总操作数（ban 10 + pick 10） */
  totalActions: number;
  /** 我方 5 人（含自己）按楼序：名字/位置/已选英雄/是否当前操作者 */
  myTeamBoard: BoardPlayer[];
  /** 对面 5 人按楼序 */
  enemyTeamBoard: BoardPlayer[];
}

/** 选人棋盘上的一个玩家 */
export interface BoardPlayer {
  cellId: number;
  summonerName: string;
  lane: string;
  championId: number;
  /** 海克斯大乱斗：该玩家翻开的全部卡（部分客户端字段形态） */
  championIds?: number[];
  isMe: boolean;
  /** 是否当前操作者（正在 ban/pick） */
  acting: boolean;
}

/** 海克斯大乱斗共享池中的英雄（翻开的英雄，带胜率/海克斯牌，标注是否自己翻的） */
export interface AramPoolEntry {
  heroId: number;
  title: string;
  alias: string;
  /** 池内推荐序号（按综合分排序，1 起；无数据英雄排最后） */
  rank: number;
  /** 百分比胜率；榜内无数据为 null */
  winRate: number | null;
  pickRate: number | null;
  /** 综合推荐分 0-100（胜率 70% + 登场率 30%）；榜外英雄为 null */
  score: number | null;
  bestAugments: { name_cn: string; winRate: number }[];
  /** 最佳搭档（组合胜率，百分比） */
  bestPartners: { title: string; winRate: number }[];
  /** 是否自己翻开（自己的卡牌） */
  isMine: boolean;
}

/** 海斗共享池英雄综合推荐分（0-100，数据驱动）：胜率 70% + 登场率 30%；无榜数据为 null */
export function aramPoolScore(winRate: number | null, pickRate: number | null): number | null {
  if (winRate === null || pickRate === null) return null;
  const wr = Math.max(0, Math.min(100, (winRate - 45) * 8)); // 胜率 45%→0 分，50%→40，55%→80
  const pr = Math.max(0, Math.min(100, pickRate * 6));       // 登场率 8%→48 分，16%→96
  return Math.round(wr * 0.7 + pr * 0.3);
}

/** 取一名玩家翻开的英雄（兼容 LCU 字段形态：championId 单值 / championIds 数组） */
function flippedHeroIds(p: { championId: number; championIds?: number[] }): number[] {
  const ids = new Set<number>();
  for (const id of p.championIds ?? []) if (id > 0) ids.add(id);
  if (p.championId > 0) ids.add(p.championId);
  return [...ids];
}

/**
 * 构建海克斯大乱斗共享池：双方已翻开（championId>0 / championIds）的英雄去重，
 * 匹配胜率榜数据（无榜数据的补空），按胜率降序。
 * 海斗共享池为 10 人共用：我方 + 对面翻开的卡都会进池，全员可选。
 */
export function buildAramPool(
  myTeam: { cellId: number; championId: number; championIds?: number[] }[],
  theirTeam: { cellId: number; championId: number; championIds?: number[] }[],
  localPlayerCellId: number,
  heroes: ReadonlyMap<number, { heroId: number; name: string; title: string; alias: string }>,
  aramHeroes: Awaited<ReturnType<typeof recommendHextechHeroes>>,
): AramPoolEntry[] {
  const flipped = [...myTeam, ...theirTeam].flatMap(flippedHeroIds);
  const mine = new Set(flippedHeroIds(myTeam.find((p) => p.cellId === localPlayerCellId) ?? { championId: 0 }));
  const rankMap = new Map(aramHeroes.map((h) => [h.heroId, h]));
  return [...new Set(flipped)]
    .map((id) => {
      const r = rankMap.get(id);
      const h = heroes.get(id);
      const winRate = r?.winRate ?? null;
      const pickRate = r?.pickRate ?? null;
      return {
        heroId: id,
        title: h?.title ?? `#${id}`,
        alias: h?.alias ?? '',
        winRate,
        pickRate,
        score: aramPoolScore(winRate, pickRate),
        bestAugments: (r?.bestAugments ?? []).map((a) => ({ name_cn: a.name_cn, winRate: a.winRate })),
        bestPartners: (r?.bestPartners ?? []).slice(0, 3).map((p) => ({ title: p.title, winRate: p.winRate })),
        isMine: mine.has(id),
      };
    })
    .sort((a, b) => (b.score ?? -1) - (a.score ?? -1)) // 有分在前按分降序，榜外（无分）排最后
    .map((entry, i) => ({ ...entry, rank: i + 1 }));
}

const TIMER_PHASE_NAMES: Record<string, string> = {
  PLANNING: '位置规划',
  BAN_PICK: '禁用/选择',
  FINALIZATION: '最终确认',
  IDLE: '等待',
};

/** 补全 summonerName（新版 LCU session 可能不含名字） */
async function enrichNames(session: ChampSelectSession): Promise<Map<number, string>> {
  const names = new Map<number, string>();
  const all = [...session.myTeam, ...session.theirTeam];
  const need = all.filter((p) => !p.summonerName && !p.isBot);
  const results = await Promise.allSettled(
    need.map((p) => getSummoner(p.summonerId).then((s) => [p.summonerId, s.displayName] as const)),
  );
  for (const r of results) {
    if (r.status === 'fulfilled') names.set(r.value[0], r.value[1]);
  }
  return names;
}

/** 读取当前选人 session 并给出 BP 建议（按模式门控：仅排位/大乱斗有推荐） */
export async function analyzeChampSelect(): Promise<ChampSelectAnalysis> {
  const session = await getChampSelectSession();
  const [heroes, extraNames, tierId] = await Promise.all([
    getHeroList(),
    enrichNames(session),
    getMyTierId().catch(() => null),
  ]);
  const tier = (tierId ?? 255) as TierId; // 查不到段位（未排位/客户端差异）用全段位
  const { mode, label } = queueToMode(session.queueId);

  const myPicks = session.myTeam.filter((p) => p.championId > 0).map((p) => p.championId);
  const enemyPicks = session.theirTeam.filter((p) => p.championId > 0).map((p) => p.championId);
  const myBans = session.bans.myTeamBans.filter((b) => b > 0);
  const enemyBans = session.bans.theirTeamBans.filter((b) => b > 0);

  // 对面英雄各自位置（assignedPosition -> 榜单 lane 口径）；缺失时 pick 服务按榜单登场率推断
  const enemyLanes: Record<number, string> = {};
  for (const p of session.theirTeam) {
    if (p.championId > 0 && p.assignedPosition && !['none', 'fill'].includes(p.assignedPosition)) {
      enemyLanes[p.championId] = normalizeLane(p.assignedPosition);
    }
  }

  const me = session.myTeam.find((p) => p.cellId === session.localPlayerCellId);
  const rawLane = me?.assignedPosition ?? '';
  const myLane = rawLane && !['none', 'fill'].includes(rawLane) ? normalizeLane(rawLane) : 'ALL';
  const myHeroId = me?.championId ?? 0;
  const myOpenLanes = session.myTeam
    .filter((p) => p.assignedPosition && !['none', 'utility', 'fill'].includes(p.assignedPosition))
    .map((p) => p.assignedPosition.toUpperCase());

  // ---------- 阶段感知：真实 BP 流程（ban 10 人同时 → pick 按楼序轮选） ----------
  const nameOf = (p: { cellId: number; summonerId: number; summonerName?: string } | undefined) => {
    if (!p) return '未知';
    return p.summonerName ?? extraNames.get(p.summonerId) ?? '召唤师';
  };
  // actions 数组按流程顺序排列（每个 cell 的 ban/pick 动作），第一个未完成的就是当前操作
  const completedActions = session.actions
    .flat()
    .filter((a) => a.completed).length;
  const totalActions = session.actions.flat().length;
  const pending = session.actions.flat().find((a) => !a.completed) ?? null;
  let currentAction: ChampSelectAnalysis['currentAction'] = null;
  if (pending) {
    const actor = [...session.myTeam, ...session.theirTeam].find((p) => p.cellId === pending.actorCellId);
    const isAlly = !!session.myTeam.find((p) => p.cellId === pending.actorCellId);
    currentAction = {
      type: pending.type === 'ban' ? 'ban' : 'pick',
      actorName: nameOf(actor),
      isAlly,
      isMe: pending.actorCellId === session.localPlayerCellId,
      lane: actor?.assignedPosition ?? '',
    };
  }
  const board = (team: typeof session.myTeam): BoardPlayer[] =>
    team
      .slice()
      .sort((a, b) => a.cellId - b.cellId) // 按楼序
      .map((p) => ({
        cellId: p.cellId,
        summonerName: nameOf(p),
        lane: p.assignedPosition ?? '',
        championId: p.championId,
        championIds: p.championIds,
        isMe: p.cellId === session.localPlayerCellId,
        acting: pending?.actorCellId === p.cellId,
      }));

  const [picks, bans, aramHeroes] = await Promise.all([
    // 仅峡谷排位激活 BP 推荐（按账号段位查对位/榜单）
    isRankedMode(mode) && enemyPicks.length
      ? recommendPick({ enemyHeroIds: enemyPicks, enemyLanes, excludeIds: myPicks, myLane, topN: 8, opts: { tier } })
      : Promise.resolve([]),
    isRankedMode(mode)
      ? recommendBan({ myHeroIds: myPicks, topN: 5, opts: { tier } }).catch(() => [])
      : Promise.resolve([]),
    // 仅海克斯大乱斗激活英雄/海克斯牌推荐
    isHextechMode(mode)
      ? recommendHextechHeroes(10)
      : Promise.resolve([]),
  ]);

  return {
    phase: session.timer.phase,
    mode,
    modeLabel: label,
    queueId: session.queueId,
    me: me?.summonerName ?? extraNames.get(me?.summonerId ?? -1) ?? '我',
    enemyPicks,
    myPicks,
    myBans,
    enemyBans,
    myOpenLanes,
    myLane,
    myHeroId,
    picks,
    bans,
    aramHeroes,
    aramPool: isHextechMode(mode) ? buildAramPool(session.myTeam, session.theirTeam, session.localPlayerCellId, heroes, aramHeroes) : [],
    tier,
    tierName: TIER_NAMES[tier] ?? '全段位',
    timerPhase: TIMER_PHASE_NAMES[session.timer.phase] ?? session.timer.phase,
    timeLeftSec: Math.max(0, Math.round((session.timer.adjustedTimeLeftInPhase ?? 0) / 1000)),
    currentAction,
    completedActions,
    totalActions,
    myTeamBoard: board(session.myTeam),
    enemyTeamBoard: board(session.theirTeam),
  };
}

File: desktop/src-tauri/src/champselect.rs
// 选人阶段分析（对应 src/services/champselect.ts）：从 LCU session 数据 -> BP 推荐

use crate::ban::{self, BanInput};
use crate::cn101::cn;
use crate::hextech::recommend_hextech_heroes;
use crate::lcu::{get_champ_select_session, get_my_tier_id, get_summoner};
use crate::models::*;
use crate::pick::{self, PickInput};
use serde_json::{json, Value};
use std::collections::HashMap;

/// 海斗共享池英雄综合推荐分（0-100）：胜率 70% + 登场率 30%
pub fn aram_pool_score(win_rate: Option<f64>, pick_rate: Option<f64>) -> Option<f64> {
    let wr = win_rate?;
    let pr = pick_rate?;
    let wr_s = (wr - 45.0).clamp(0.0, 100.0) * 8.0;
    let pr_s = pr.clamp(0.0, 100.0) * 6.0;
    Some(((wr_s * 0.7 + pr_s * 0.3) * 100.0).round() / 100.0)
}

fn timer_phase_name(phase: &str) -> String {
    match phase {
        "PLANNING" => "位置规划".to_string(),
        "BAN_PICK" => "禁用/选择".to_string(),
        "FINALIZATION" => "最终确认".to_string(),
        "IDLE" => "等待".to_string(),
        _ => phase.to_string(),
    }
}

/// 读取当前选人 session 并给出 BP 建议
pub async fn analyze_champ_select() -> Result<Value, String> {
    let session = get_champ_select_session().await?;
    let heroes = cn().get_hero_list(false).await?;
    let tier_id = get_my_tier_id().await;
    let tier: TierId = tier_id.unwrap_or(255);
    let queue_id = session.get("queueId").and_then(|v| v.as_i64()).unwrap_or(0);
    let (mode, label) = queue_to_mode(queue_id);

    let my_team = session.get("myTeam").and_then(|v| v.as_array()).cloned().unwrap_or_default();
    let their_team = session.get("theirTeam").and_then(|v| v.as_array()).cloned().unwrap_or_default();
    let local_cell = session.get("localPlayerCellId").and_then(|v| v.as_i64()).unwrap_or(-1);

    // 补全 summonerName（新版 LCU session 可能不含名字）
    let mut extra_names: HashMap<i64, String> = HashMap::new();
    for p in my_team.iter().chain(their_team.iter()) {
        let has_name = p.get("summonerName").and_then(|v| v.as_str()).map(|s| !s.is_empty()).unwrap_or(false);
        if has_name {
            continue;
        }
        let is_bot = p.get("isBot").and_then(|v| v.as_bool()).unwrap_or(false);
        if is_bot {
            continue;
        }
        if let Some(sid) = p.get("summonerId").and_then(|v| v.as_i64()) {
            if let Ok(s) = get_summoner(sid).await {
                let name = crate::lcu::summoner_display_name(&s);
                if !name.is_empty() {
                    extra_names.insert(sid, name);
                }
            }
        }
    }

    let cid = |p: &Value| p.get("championId").and_then(|v| v.as_i64()).unwrap_or(0);
    let my_picks: Vec<i64> = my_team.iter().filter(|p| cid(p) > 0).map(cid).collect();
    let enemy_picks: Vec<i64> = their_team.iter().filter(|p| cid(p) > 0).map(cid).collect();
    let bans = session.get("bans").cloned().unwrap_or(Value::Null);
    let my_bans: Vec<i64> = bans.get("myTeamBans").and_then(|v| v.as_array()).unwrap_or(&vec![])
        .iter().filter(|b| b.as_i64().unwrap_or(0) > 0).filter_map(|b| b.as_i64()).collect();
    let enemy_bans: Vec<i64> = bans.get("theirTeamBans").and_then(|v| v.as_array()).unwrap_or(&vec![])
        .iter().filter(|b| b.as_i64().unwrap_or(0) > 0).filter_map(|b| b.as_i64()).collect();

    // 对面英雄各自位置
    let mut enemy_lanes: HashMap<i64, String> = HashMap::new();
    for p in &their_team {
        let hero_id = cid(p);
        if hero_id <= 0 {
            continue;
        }
        let pos = p.get("assignedPosition").and_then(|v| v.as_str()).unwrap_or("");
        if !pos.is_empty() && pos != "none" && pos != "fill" {
            enemy_lanes.insert(hero_id, pick::normalize_lane(pos));
        }
    }

    let me = my_team.iter().find(|p| p.get("cellId").and_then(|v| v.as_i64()).unwrap_or(-1) == local_cell).cloned();
    let raw_lane = me.as_ref().and_then(|m| m.get("assignedPosition")).and_then(|v| v.as_str()).unwrap_or("");
    let my_lane = if raw_lane.is_empty() || raw_lane == "none" || raw_lane == "fill" {
        "ALL".to_string()
    } else {
        pick::normalize_lane(raw_lane)
    };
    let my_hero_id = me.as_ref().map(&cid).unwrap_or(0);
    let my_open_lanes: Vec<String> = my_team.iter()
        .filter(|p| {
            let pos = p.get("assignedPosition").and_then(|v| v.as_str()).unwrap_or("");
            !pos.is_empty() && pos != "none" && pos != "utility" && pos != "fill"
        })
        .map(|p| p.get("assignedPosition").and_then(|v| v.as_str()).unwrap_or("").to_uppercase())
        .collect();

    // ---------- 阶段感知 ----------
    let name_of = |p: Option<&Value>| -> String {
        match p {
            Some(p) => p.get("summonerName").and_then(|v| v.as_str()).filter(|s| !s.is_empty())
                .map(String::from)
                .or_else(|| p.get("summonerId").and_then(|v| v.as_i64()).and_then(|id| extra_names.get(&id).cloned()))
                .unwrap_or_else(|| "召唤师".to_string()),
            None => "未知".to_string(),
        }
    };
    let actions: Vec<Value> = session.get("actions").and_then(|v| v.as_array()).cloned().unwrap_or_default();
    let flat: Vec<&Value> = actions.iter().flat_map(|a| {
        a.as_array().map(|arr| arr.iter().collect::<Vec<&Value>>()).unwrap_or_default()
    }).collect();
    let completed = flat.iter().filter(|a| a.get("completed").and_then(|v| v.as_bool()).unwrap_or(false)).count();
    let total = flat.len();
    let pending = flat.iter().find(|a| !a.get("completed").and_then(|v| v.as_bool()).unwrap_or(false)).copied();
    let mut current_action: Option<Value> = None;
    if let Some(p) = pending {
        let actor_cell = p.get("actorCellId").and_then(|v| v.as_i64()).unwrap_or(-1);
        let all: Vec<&Value> = my_team.iter().chain(their_team.iter()).collect();
        let actor = all.iter().find(|x| x.get("cellId").and_then(|v| v.as_i64()).unwrap_or(-1) == actor_cell).copied();
        let is_ally = my_team.iter().any(|x| x.get("cellId").and_then(|v| v.as_i64()).unwrap_or(-1) == actor_cell);
        let atype = p.get("type").and_then(|v| v.as_str()).unwrap_or("");
        current_action = Some(json!({
            "type": if atype == "ban" { "ban" } else { "pick" },
            "actorName": name_of(actor),
            "isAlly": is_ally,
            "isMe": actor_cell == local_cell,
            "lane": actor.and_then(|a| a.get("assignedPosition")).and_then(|v| v.as_str()).unwrap_or(""),
        }));
    }
    let board = |team: &Vec<Value>| -> Vec<Value> {
        let mut sorted: Vec<&Value> = team.iter().collect();
        sorted.sort_by(|a, b| {
            a.get("cellId").and_then(|v| v.as_i64()).unwrap_or(0)
                .cmp(&b.get("cellId").and_then(|v| v.as_i64()).unwrap_or(0))
        });
        sorted.into_iter().map(|p| {
            let cell = p.get("cellId").and_then(|v| v.as_i64()).unwrap_or(-1);
            json!({
                "cellId": cell,
                "summonerName": name_of(Some(p)),
                "lane": p.get("assignedPosition").and_then(|v| v.as_str()).unwrap_or(""),
                "championId": cid(p),
                "isMe": cell == local_cell,
                "acting": pending.map(|a| a.get("actorCellId").and_then(|v| v.as_i64()).unwrap_or(-1) == cell).unwrap_or(false),
            })
        }).collect()
    };
    let my_board = board(&my_team);
    let enemy_board = board(&their_team);

    // ---------- 推荐 ----------
    let ranked = is_ranked_mode(mode);
    let hex = is_hextech_mode(mode);
    let picks: Vec<Value> = if ranked && !enemy_picks.is_empty() {
        let input = PickInput {
            enemy_hero_names: Vec::new(),
            enemy_hero_ids: enemy_picks.clone(),
            enemy_lanes: enemy_lanes.clone(),
            my_lane: my_lane.clone(),
            exclude: Vec::new(),
            exclude_ids: my_picks.clone(),
            top_n: 8,
            tier,
        };
        pick::recommend_pick(&input).await.unwrap_or_default().into_iter().map(|r| serde_json::to_value(r).unwrap_or(Value::Null)).collect()
    } else {
        Vec::new()
    };
    let bans: Vec<Value> = if ranked {
        let input = BanInput { my_hero_names: Vec::new(), my_hero_ids: my_picks.clone(), top_n: 5, tier };
        ban::recommend_ban(&input).await.unwrap_or_default().into_iter().map(|r| serde_json::to_value(r).unwrap_or(Value::Null)).collect()
    } else {
        Vec::new()
    };
    let aram_heroes: Vec<Value> = if hex {
        recommend_hextech_heroes(10).await.unwrap_or_default().into_iter().map(|h| serde_json::to_value(h).unwrap_or(Value::Null)).collect()
    } else {
        Vec::new()
    };

    // 海斗共享池
    let aram_pool: Vec<Value> = if hex {
        let bench = session.get("benchChampions").and_then(|v| v.as_array()).cloned().unwrap_or_default();
        build_aram_pool(&bench, &heroes, &aram_heroes)
    } else {
        Vec::new()
    };

    // 我的英雄海斗榜推荐
    let my_hero_stat: Option<Value> = if my_hero_id > 0 {
        recommend_hextech_heroes(300).await.ok()
            .and_then(|list| list.into_iter().find(|h| h.hero_id == my_hero_id))
            .map(|h| json!({
                "title": h.title,
                "alias": h.alias,
                "winRate": h.win_rate,
                "pickRate": h.pick_rate,
                "rank": h.rank,
                "bestAugments": h.best_augments.iter().take(5).map(|a| json!({ "name_cn": a.name_cn, "winRate": a.win_rate, "pickRate": a.pick_rate })).collect::<Vec<_>>(),
                "bestPartners": h.best_partners.iter().take(3).map(|p| json!({ "title": p.title, "winRate": p.win_rate })).collect::<Vec<_>>(),
            }))
    } else {
        None
    };

    let timer = session.get("timer").cloned().unwrap_or(Value::Null);
    let phase = timer.get("phase").and_then(|v| v.as_str()).unwrap_or("").to_string();
    let time_left = timer.get("adjustedTimeLeftInPhase").and_then(|v| v.as_i64()).unwrap_or(0).max(0);

    let me_name = me.as_ref()
        .and_then(|m| m.get("summonerName").and_then(|v| v.as_str()).filter(|s| !s.is_empty()).map(String::from))
        .or_else(|| me.as_ref().and_then(|m| m.get("summonerId").and_then(|v| v.as_i64())).and_then(|id| extra_names.get(&id).cloned()))
        .unwrap_or_else(|| "我".to_string());

    Ok(json!({
        "phase": phase,
        "mode": mode.as_str(),
        "modeLabel": label,
        "queueId": queue_id,
        "me": me_name,
        "enemyPicks": enemy_picks,
        "myPicks": my_picks,
        "myBans": my_bans,
        "enemyBans": enemy_bans,
        "myOpenLanes": my_open_lanes,
        "myLane": my_lane,
        "myHeroId": my_hero_id,
        "myHeroStat": my_hero_stat,
        "picks": picks,
        "bans": bans,
        "aramHeroes": aram_heroes,
        "aramPool": aram_pool,
        "tier": tier,
        "tierName": tier_names(tier),
        "timerPhase": timer_phase_name(&phase),        "timeLeftSec": time_left,
        "currentAction": current_action,
        "completedActions": completed,
        "totalActions": total,
        "myTeamBoard": my_board,
        "enemyTeamBoard": enemy_board,
    }))
}

/// 构建海克斯大乱斗共享池（只读 LCU session.benchChampions）
pub fn build_aram_pool(
    bench_champions: &[Value],
    heroes: &HashMap<i64, ChampionBase>,
    aram_heroes: &[Value],
) -> Vec<Value> {
    let mut ids: Vec<i64> = bench_champions.iter()
        .filter_map(|b| b.get("championId").and_then(|v| v.as_i64()))
        .filter(|id| *id > 0)
        .collect::<std::collections::HashSet<_>>()
        .into_iter().collect();
    ids.sort_unstable();
    let rank_map: HashMap<i64, &Value> = aram_heroes.iter()
        .filter_map(|h| h.get("heroId").and_then(|v| v.as_i64()).map(|id| (id, h)))
        .collect();
    let mut entries: Vec<Value> = ids.into_iter().map(|id| {
        let r = rank_map.get(&id).copied();
        let h = heroes.get(&id);
        let win_rate = r.and_then(|r| r.get("winRate").and_then(|v| v.as_f64()));
        let pick_rate = r.and_then(|r| r.get("pickRate").and_then(|v| v.as_f64()));
        let best_augments: Vec<Value> = r.and_then(|r| r.get("bestAugments").and_then(|v| v.as_array())).unwrap_or(&vec![])
            .iter().map(|a| json!({ "name_cn": a.get("name_cn").and_then(|v| v.as_str()).unwrap_or(""), "winRate": a.get("winRate").and_then(|v| v.as_f64()).unwrap_or(0.0) })).collect();
        let best_partners: Vec<Value> = r.and_then(|r| r.get("bestPartners").and_then(|v| v.as_array())).unwrap_or(&vec![])
            .iter().take(3).map(|p| json!({ "title": p.get("title").and_then(|v| v.as_str()).unwrap_or(""), "winRate": p.get("winRate").and_then(|v| v.as_f64()).unwrap_or(0.0) })).collect();
        json!({
            "heroId": id,
            "title": h.map(|x| x.title.clone()).unwrap_or_else(|| format!("#{id}")),
            "alias": h.map(|x| x.alias.clone()).unwrap_or_default(),
            "winRate": win_rate,
            "pickRate": pick_rate,
            "score": aram_pool_score(win_rate, pick_rate),
            "bestAugments": best_augments,
            "bestPartners": best_partners,
        })
    }).collect();
    entries.sort_by(|a, b| {
        let sa = a.get("score").and_then(|v| v.as_f64()).unwrap_or(-1.0);
        let sb = b.get("score").and_then(|v| v.as_f64()).unwrap_or(-1.0);
        sb.partial_cmp(&sa).unwrap_or(std::cmp::Ordering::Equal)
    });
    for (i, e) in entries.iter_mut().enumerate() {
        e["rank"] = json!(i + 1);
    }
    entries
}
